### PR TITLE
[RHCLOUD-48478] Lock roles during assignment and seeding

### PR DESCRIPTION
**Pull request by randomnetcat**
## Link(s) to Jira
[RHCLOUD-48478](https://redhat.atlassian.net/browse/RHCLOUD-48478)

## Description of Intent of Change(s)
Lock system roles `FOR UPDATE` during seeding and `FOR SHARE` during assignment. This ensures that, once seeding has finished, any further assignments will use the new permissions to compute the scope.

(As described in the comment in `definer.py`, this doesn't handle cases where the scope settings have changed, but it's still an improvement.)

~~Do not merge until #3058 is merged, since otherwise we will potentially be holding locks on system roles while updating their binding scope (which can take a while).~~

[RHCLOUD-48478]: https://redhat.atlassian.net/browse/RHCLOUD-48478?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved consistency when assigning, removing, and resetting roles across access and group relationships.
  - Strengthened handling of system roles during concurrent updates to prevent conflicting changes.
  - Improved role creation and updates during setup, reducing the risk of duplicate or inconsistent role data.

- **Refactor**
  - Role processing now supports broader collections and preserves the supplied processing order.
  - Added safeguards to detect missing roles during relationship updates.

### CODE REVIEW

**Comment by lpichler** [comment]
## Dimensional Review -- Action Required

- **high** `rbac/management/role/definer.py:159` -- No test coverage for new locking behavior — the core purpose of this PR is untested *(test-quality)*

<details><summary>Full report</summary>

# Dimensional Review: PR #3059

**PR**: [RHCLOUD-48478: Lock roles during assignment and seeding](https://github.com/project-kessel/insights-rbac/pull/3059)
**Author**: randomnetcat (Janet Cobb) | **Branch**: role-seed-locking -> master
**Changes**: +75 -10 across 4 files
**Dimensions**: 13 dispatched (3 general, 9 specialist, 1 meta)
**Models**: 4 opus, 8 sonnet, 1 sonnet (medium effort)

## Summary

- **12 unique findings** (from 42 raw across 13 dimensions)
- **Severity**: 0 critical, 1 high, 7 medium, 4 low
- **3 consensus findings** (agreed by 3+ dimensions)
- **1 disputed finding** (db-query flagged critical, concurrency refuted)

## Consensus Findings

### 1. select_for_update().get_or_create() in SERIALIZABLE transaction has complex semantics
**Severity**: medium | **File**: `rbac/management/role/definer.py:159` | **Agreed by**: 6 dimensions

> `Role.objects.select_for_update().get_or_create()` within `@atomic_with_retry(retries=3)` creates a complex three-layer scheme: SERIALIZABLE isolation + explicit FOR UPDATE + retry logic. The FOR UPDATE only applies to existing rows (not creates). Multiple dimensions flagged the overlapping mechanisms and the acknowledged TODO about settings-based scope changes.

**Suggestion**: Add a comment clarifying the cross-transaction coordination purpose. Consider splitting get() with select_for_update() from create() for clarity. Track the TODO as a separate issue.

**Alternatives**:
1. **Split into try get() with select_for_update(), except DoesNotExist: create()** -- Clearer intent but more verbose; handles concurrent creates via retry **(recommended)**
2. **Keep current pattern with comprehensive comment** -- Minimal code change but relies on documentation
3. **Use try create() first, on IntegrityError do select_for_update().get()** -- Avoids unnecessary locking on first seed, but more complex error handling

*Dimensions*: code-quality, architecture, db-query, integration, error-handling, concurrency
*Source skills*: agentic-sdlc:pr-review-detailed, agentic-sdlc:pr-review, db-query-reviewer, integration-reviewer, error-handling-reviewer, concurrency-reviewer

---

### 2. Raw SQL FOR SHARE pattern inconsistent with established codebase pattern
**Severity**: medium | **File**: `rbac/management/group/relation_api_dual_write_subject_handler.py:120` | **Agreed by**: 4 dimensions

> The new `_with_system_roles_for_share` uses `Role.objects.raw()` with f-string interpolation for the IN clause. The codebase already has a FOR SHARE pattern in `management/tenant_mapping/v2_activation.py` using `django.db.connection.cursor()` with static SQL constants. Two different patterns now exist for the same purpose.

**Suggestion**: Follow the v2_activation.py pattern or extract a reusable helper. Add a comment explaining why raw SQL is needed (Django ORM has no `select_for_share()`).

**Alternatives**:
1. **Follow v2_activation.py pattern with connection.cursor() and static SQL constant** -- Consistent with codebase **(recommended)**
2. **Extract a reusable _for_share() helper method** -- Eliminates future inconsistency but adds abstraction
3. **Keep current objects.raw() with added comment** -- Minimal change, inconsistency persists

*Dimensions*: code-quality, architecture, security, integration
*Source skills*: agentic-sdlc:pr-review-detailed, agentic-sdlc:pr-review, security-reviewer, integration-reviewer

---

### 3. RuntimeError with internal integer IDs should use domain exception
**Severity**: medium | **File**: `rbac/management/group/relation_api_dual_write_subject_handler.py:133` | **Agreed by**: 3 dimensions

> `raise RuntimeError(f"Could not find roles with the following IDs: {requested_ids - final_ids}")` uses generic RuntimeError instead of a domain exception, and includes internal integer role IDs. Per project security guidelines, integer PKs should not be exposed. While this won't reach API responses, IDs appear in server logs.

**Suggestion**: Use a domain-specific exception and role UUIDs/names instead of integer IDs. Log internal IDs separately for debugging.

**Alternatives**:
1. **Domain exception with role UUIDs, log integer IDs separately** -- Consistent with project conventions **(recommended)**
2. **Keep RuntimeError but remove integer IDs** -- Quick fix but wrong exception type
3. **Use Django's ObjectDoesNotExist** -- ORM-consistent but may be caught by unrelated handlers

*Dimensions*: error-handling, security, db-query
*Source skills*: error-handling-reviewer, security-reviewer, db-query-reviewer

---

## High Severity Findings

### 4. No test coverage for new locking behavior
**Severity**: high | **File**: `rbac/management/role/definer.py:159` + `rbac/management/group/relation_api_dual_write_subject_handler.py:96`

> The core purpose of this PR (preventing race conditions via locking) has no tests. The new `_with_system_roles_for_share` method (40+ lines including raw SQL) is completely untested. No concurrency tests verify FOR SHARE blocks concurrent FOR UPDATE. No edge case tests for empty lists, mixed roles, or the RuntimeError path.

**Suggestion**: Add TransactionalIdentityRequest tests: (1) unit tests for `_with_system_roles_for_share` edge cases, (2) concurrency tests verifying lock coordination.

*Dimension*: test-quality | *Source*: test-reviewer

---

## Medium Severity Findings

### 5. Implicit locking contract -- selective calling creates fragile rules
**Severity**: medium | **File**: `rbac/management/group/relation_api_dual_write_subject_handler.py:461`

> `_with_system_roles_for_share` is called in some paths but not others, with the rationale scattered across 4 comment blocks in 2 files. A future developer adding a new V1 assignment path could easily miss the requirement.

**Suggestion**: Extend the docstring to explicitly document which patterns require the lock and which don't.

*Dimension*: architecture | *Source*: agentic-sdlc:pr-review

---

### 6. Raw SQL lacks database exception handling
**Severity**: medium | **File**: `rbac/management/group/relation_api_dual_write_subject_handler.py:120`

> `Role.objects.raw()` used without try-except. Can raise OperationalError (deadlocks, lock timeouts) that propagate as unhandled 500s.

**Suggestion**: Wrap in try-except for OperationalError to provide meaningful context before re-raising.

*Dimension*: error-handling | *Source*: error-handling-reviewer

---

### 7. Unbounded IN clause in raw SQL query
**Severity**: medium | **File**: `rbac/management/group/relation_api_dual_write_subject_handler.py:117`

> No limit check on the number of role IDs passed to the IN clause. While system roles are platform-defined and finite, a defensive check prevents potential resource exhaustion.

**Suggestion**: Add `if len(system) > 1000: raise ValueError('Too many roles to lock')`.

*Dimension*: vulnerability | *Source*: vuln-reviewer

---

### 8. TODO acknowledges unresolved settings-based race condition
**Severity**: medium | **File**: `rbac/management/role/definer.py:157`

> The locking only partially solves the race condition. Settings-based scope changes (not permission-based) can still cause stale scope calculations across processes.

**Suggestion**: File a follow-up ticket. Document limitation in operational runbooks.

*Dimensions*: integration, architecture | *Sources*: integration-reviewer, agentic-sdlc:pr-review

---

## Low Severity Findings

### 9. FOR SHARE query lacks ORDER BY for deterministic lock acquisition
**Severity**: low | **File**: `rbac/management/group/relation_api_dual_write_subject_handler.py:121`

> Currently safe (FOR SHARE locks are compatible), but adding `ORDER BY id` guards against future changes.

*Dimension*: concurrency | *Source*: concurrency-reviewer

---

### 10. Non-idiomatic Python: `if len(system) > 0:`
**Severity**: low | **File**: `rbac/management/group/relation_api_dual_write_subject_handler.py:116`

> Should be `if system:` per Python idioms.

*Dimension*: code-quality | *Source*: agentic-sdlc:pr-review-detailed

---

### 11. DISPUTED: _add_car_roles_v1 allegedly lacks transaction wrapper
**Severity**: low (downgraded from critical) | **File**: `rbac/api/cross_access/relation_api_dual_write_cross_access_handler.py:12`

> db-query dimension flagged this as **critical**, claiming FOR SHARE locks would be released immediately. However, **concurrency dimension refuted this** by tracing the full call chain: the `partial_update` view is decorated with `@atomic (SERIALIZABLE)`, providing the transaction context. No code change needed.

*Dimension*: db-query (disputed by concurrency) | *Source*: db-query-reviewer

---

### 12. Missing return type annotation
**Severity**: low | **File**: `rbac/management/group/relation_api_dual_write_subject_handler.py:96`

> `_with_system_roles_for_share` has parameter type hints but no return type annotation.

**Suggestion**: Add `-> list[Role]:` to the signature.

*Dimension*: code-quality | *Source*: agentic-sdlc:pr-review-detailed

---

## Per-Dimension Summaries

### correctness (general, opus) -- PASS | source: productivity-utils:code-review (LIVE)
0 findings
> No correctness issues found. Locking logic, SQL parameterization, type handling, and control flow are all correct.

### code-quality (general, sonnet) -- PASS | source: agentic-sdlc:pr-review-detailed (STALE)
5 findings (1 high, 1 medium, 3 low)
> Generally good code quality. Minor style and annotation issues. Raw SQL concern overlaps with architecture.

### architecture (general, opus) -- WARN | source: agentic-sdlc:pr-review (LIVE)
3 findings (3 medium)
> Sound locking approach with three concerns: inconsistent raw SQL pattern, implicit locking contract, and complex SERIALIZABLE + FOR UPDATE interaction.

### security (specialist, opus) -- PASS | source: security-reviewer (LIVE)
3 findings (2 info, 1 low)
> This PR is a security improvement. Raw SQL is safe (parameterized). Minor concern about integer IDs in error message.

### performance (specialist, sonnet) -- WARN | source: performance-reviewer (LIVE)
4 findings (1 medium, 2 low, 1 info)
> Minor inefficiencies in _with_system_roles_for_share (unnecessary set conversion, redundant iteration). Negligible for typical role counts.

### error-handling (specialist, sonnet) -- WARN | source: error-handling-reviewer (LIVE)
3 findings (2 medium, 1 low)
> Missing exception handling around raw SQL, RuntimeError instead of domain exception.

### test-quality (specialist, sonnet) -- WARN | source: test-reviewer (LIVE)
6 findings (2 high, 3 medium, 1 low)
> The core locking behavior is untested. No concurrency tests, no edge case tests for the new method.

### concurrency (specialist, opus) -- PASS | source: concurrency-reviewer (LIVE)
5 findings (3 info, 2 low)
> Locking strategy is sound. FOR SHARE locks are compatible (no deadlock). FOR UPDATE during seeding correctly blocks concurrent assignments. Refuted the db-query dimension's critical finding about missing transaction boundaries.

### api-contract (specialist, sonnet) -- PASS | source: api-contract-reviewer (LIVE)
0 findings
> No API contract impact. All changes are to internal private methods. Type signature widening (set->Iterable) is backward compatible.

### db-query (specialist, sonnet) -- FAIL -> WARN | source: db-query-reviewer (LIVE)
5 findings (1 critical-disputed, 1 medium, 3 low)
> The critical finding (missing transaction wrapper) was refuted by the concurrency dimension. Remaining findings about SQL style and error handling overlap with other dimensions.

### integration (specialist, sonnet) -- WARN | source: integration-reviewer (LIVE)
6 findings (3 medium, 3 low)
> Concerns about transaction boundaries (refuted), inconsistent v1/v2 locking strategy (intentional), and unresolved settings-based race condition.

### vulnerability (specialist, sonnet) -- WARN | source: vuln-reviewer (STALE)
2 findings (1 medium, 1 low)
> Unbounded IN clause and lock timeout concerns. No SQL injection risk (parameterized queries).

### learning-extraction (meta, sonnet) -- PASS | source: productivity-utils:learn-from-reviews (LIVE)
8 patterns extracted
> Good patterns: defensive locking, consistent lock ordering, documenting where locks are NOT needed, separate handling of system vs custom roles.

---

## All Findings

| # | Sev | File | Finding | Dims | Source |
|---|-----|------|---------|------|--------|
| 1 | medium | subject_handler.py:120 | Raw SQL FOR SHARE inconsistent with codebase pattern | 4 | agentic-sdlc:pr-review |
| 2 | medium | definer.py:159 | select_for_update().get_or_create() complex semantics | 6 | agentic-sdlc:pr-review |
| 3 | medium | subject_handler.py:133 | RuntimeError with internal integer IDs | 3 | error-handling-reviewer |
| 4 | high | definer.py:159 + subject_handler.py:96 | No test coverage for locking behavior | 1 | test-reviewer |
| 5 | medium | subject_handler.py:461 | Implicit locking contract | 1 | agentic-sdlc:pr-review |
| 6 | medium | subject_handler.py:120 | Raw SQL lacks exception handling | 1 | error-handling-reviewer |
| 7 | medium | subject_handler.py:117 | Unbounded IN clause | 1 | vuln-reviewer |
| 8 | medium | definer.py:157 | TODO: settings-based race condition unresolved | 2 | integration-reviewer |
| 9 | low | subject_handler.py:121 | FOR SHARE lacks ORDER BY | 1 | concurrency-reviewer |
| 10 | low | subject_handler.py:116 | Non-idiomatic len() check | 1 | agentic-sdlc:pr-review-detailed |
| 11 | low | cross_access_handler.py:12 | DISPUTED: missing transaction (refuted) | 1 | db-query-reviewer |
| 12 | low | subject_handler.py:96 | Missing return type annotation | 1 | agentic-sdlc:pr-review-detailed |

**Tokens**: 628,026 | **Est. cost**: ~$9.42


</details>

**Comment by lpichler** [comment]
## Code Review Findings

**File:** `rbac/management/group/relation_api_dual_write_subject_handler.py` (line 120)

**medium**: Raw SQL FOR SHARE pattern inconsistent with established codebase pattern (v2_activation.py uses connection.cursor())

The new _with_system_roles_for_share method uses Role.objects.raw() with f-string interpolation for the IN clause. The codebase already has a FOR SHARE pattern in management/tenant_mapping/v2_activation.py using django.db.connection.cursor() with static SQL constants. Django ORM has no select_for_share() support, making raw SQL necessary, but two different patterns exist for the same purpose.



---

**File:** `rbac/management/group/relation_api_dual_write_subject_handler.py` (line 133)

**medium**: RuntimeError with internal integer IDs should use domain exception and avoid exposing internal PKs

raise RuntimeError(f'Could not find roles with the following IDs: {requested_ids - final_ids}') — uses generic RuntimeError instead of a domain exception, and includes internal integer role IDs which per project security guidelines should not be exposed. While this won't reach API responses (Django returns generic 500 in production), the IDs appear in server logs.



---

**File:** `rbac/management/role/definer.py` (line 159)

**high**: No test coverage for new locking behavior — the core purpose of this PR is untested

The PR adds SELECT FOR UPDATE during seeding and FOR SHARE during assignment to prevent race conditions, but no concurrency tests verify this works. The _with_system_roles_for_share method (40+ lines of new code including raw SQL) has no tests at all. No tests verify: empty role lists, mixed system/custom roles, RuntimeError on missing roles, concurrent seeding+assignment, or deadlock prevention.



---

**File:** `rbac/management/group/relation_api_dual_write_subject_handler.py` (line 120)

**medium**: Raw SQL query does not handle database exceptions (OperationalError for deadlocks, lock timeouts)

Role.objects.raw() is used without try-except. Raw SQL can raise OperationalError (deadlocks, lock timeouts), DatabaseError, or ProgrammingError. These propagate as unhandled 500 errors without meaningful context.



---

**File:** `rbac/management/group/relation_api_dual_write_subject_handler.py` (line 117)

**medium**: Unbounded IN clause in raw SQL query — no limit on number of role IDs

id_params = ', '.join(['%s'] * len(system)) — the number of parameters comes from the roles iterable without bounds checking. While values are properly parameterized (no SQL injection), an excessively large number of role IDs could cause resource exhaustion.



---

**File:** `rbac/management/group/relation_api_dual_write_subject_handler.py` (line 121)

**low**: FOR SHARE query lacks ORDER BY for deterministic lock acquisition

SELECT * FROM management_role WHERE system = TRUE AND id IN (...) FOR SHARE — no ORDER BY clause. Currently safe because FOR SHARE locks are compatible (concurrent assignments don't deadlock), but adding ORDER BY id would guard against future changes that might introduce FOR UPDATE on this path.



---

**File:** `rbac/management/group/relation_api_dual_write_subject_handler.py` (line 116)

**low**: Non-idiomatic Python: 'if len(system) > 0:' should be 'if system:'

Line 116 uses 'if len(system) > 0:' which is less Pythonic. The idiomatic way to check if a list is non-empty is 'if system:'.



---

**File:** `rbac/management/group/relation_api_dual_write_subject_handler.py` (line 96)

**low**: Missing return type annotation on _with_system_roles_for_share

The method has parameter type hints (roles: Iterable[Role]) but lacks the return type annotation -> list[Role].

**Comment by vbelchio** [comment]
QE: NA

File: rbac/management/group/relation_api_dual_write_subject_handler.py [hunk 1]
```diff
@@ -459,3 +459,46 @@ def _migrate_custom_role(self, role):
...
+                system.append(role)
+            else:
+                custom.append(role)
+
+        if len(system) > 0:
+            id_params = ", ".join(["%s"] * len(system))
+
+            locked_system = list(
+                Role.objects.raw(
+                    f"SELECT * FROM management_role WHERE system = TRUE AND id IN ({id_params}) FOR SHARE",
```
Comment (astrozzc) [inline rbac/management/group/relation_api_dual_write_subject_handler.py:511]: For FOR_SHARE, let's use ORDER_BY to avoid deadlock because seeding sorts roles by name.

**Review by astrozzc — APPROVED** [review]

File: rbac/management/group/relation_api_dual_write_subject_handler.py (hunk 1 above)
Comment (randomnetcat) [reply rbac/management/group/relation_api_dual_write_subject_handler.py:511]: Ah, that sorting doesn't actually do anything anymore, since each role is updated in a separate transaction. I'll just remove it.

**Comment by randomnetcat** [comment]
Libor:

> No test coverage for new locking behavior — the core purpose of this PR is untested (test-quality)

It's an implementation detail, and it's tested by virtue of being used everywhere dual-write is. We don't have a good way to test the locking semantics themselves. (That might be good to develop, but we don't do it anywhere else, so it's definitely out of scope of this PR.) 

> 1. select_for_update().get_or_create() in SERIALIZABLE transaction has complex semantics

I think the comment is sufficient here? `select_for_update().get_or_create()` is an idiom already used a couple times elsewhere in the codebase.

> 2. Raw SQL FOR SHARE pattern inconsistent with established codebase pattern

We can't hardcode the query without loading the roles one at a time, which we shouldn't do for performance reasons. I also don't think we can meaningfully add a generic `_for_share()` method without a *lot* of faff based on how `FOR UPDATE` appears to be implemented in Django. (We could potentially contribute it upstream, but that might not be worth the hassle.)

> 3. RuntimeError with internal integer IDs should use domain exception

This is closer to an assertion failure than anything else; it would fire if a system role has unexpectedly vanished.

> 4. No test coverage for new locking behavior

Addressed above.

> 5. Implicit locking contract -- selective calling creates fragile rules

Fixed.

> 6. Raw SQL lacks database exception handling

So do most other database operations we do? The raw query doesn't make it special.

> 7. Unbounded IN clause in raw SQL query

Only a finite number of system roles exist, and this only operations on role objects that have already been found, so we'll never find more system roles than actually exist.

> 8. TODO acknowledges unresolved settings-based race condition

[RHCLOUD-48950](https://redhat.atlassian.net/browse/RHCLOUD-48950).

> 9. FOR SHARE query lacks ORDER BY for deterministic lock acquisition

We're inconsistent everywhere about sorting before locking. If this is actually a concern, we'd need to do something more systematic, since adding it in one place won't actually fix anything.

[RHCLOUD-48950]: https://redhat.atlassian.net/browse/RHCLOUD-48950?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
